topdown/arithmetic.go

```diff
@@ -54,6 +62,68 @@ func arithFloor(a *big.Float) (*big.Float, error) {
 	return new(big.Float).Sub(f, big.NewFloat(1.0)), nil
 }
 
+func builtinPlus(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	n1, err := builtins.NumberOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+	n2, err := builtins.NumberOperand(operands[1].Value, 2)
+	if err != nil {
+		return err
+	}
+
+	x, ok1 := n1.Int()
+	y, ok2 := n2.Int()
+
+	if ok1 && ok2 {
+		if _, ok := smallNumCache[x]; ok {
+			if _, ok := smallNumCache[y]; ok {
+				z := x + y
+				if _, ok := smallNumCache[z]; ok {
+					return iter(ast.IntNumberTerm(z))
+				}
+			}
+		}
+	}
+
+	f, err := arithPlus(builtins.NumberToFloat(n1), builtins.NumberToFloat(n2))
+	if err != nil {
+		return err
+	}
+	return iter(ast.NewTerm(builtins.FloatToNumber(f)))
+}
+
+func builtinMultiply(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	n1, err := builtins.NumberOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+	n2, err := builtins.NumberOperand(operands[1].Value, 2)
+	if err != nil {
+		return err
+	}
+
+	x, ok1 := n1.Int()
+	y, ok2 := n2.Int()
+
+	if ok1 && ok2 {
+		if _, ok := smallNumCache[x]; ok {
+			if _, ok := smallNumCache[y]; ok {
```

Comment:
I imagine we'd want to cache "big" numbers (i.e. the big number types, not necessarily big numbers) but for small ints, I guess we could just do something like:

```
if ok1 && ok2 && x < 256 && y < 256 {
```

Doesn't matter much though, and I'm happy to see this improvement! 😃